pkg/trace/writer/sender.go

```diff
@@ -1,383 +1,395 @@
-	// If MaxAge <= 0 then age limitation is disabled so do nothing
-	if s.conf.MaxAge <= 0 {
+// recordEvent records the occurrence of the given event type t. It additionally
+// passes on the data and augments it with additional information.
+func (s *sender) recordEvent(t eventType, data *eventData) {
+	if s.cfg.recorder == nil {
 		return
 	}
+	data.host = s.cfg.url.Hostname()
+	data.connectionFill = float64(len(s.climit)) / float64(cap(s.climit))
+	data.queueFill = float64(len(s.queue)) / float64(cap(s.queue))
+	s.cfg.recorder.recordEvent(t, data)
+}
 
-	var next *list.Element
+// userAgent is the computed user agent we'll use when communicating with Datadog
+var userAgent = fmt.Sprintf("Datadog Trace Agent/%s/%s", info.Version, info.GitCommit)
 
-	for e := s.queuedPayloads.Front(); e != nil; e = next {
-		payload := e.Value.(*payload)
+// retriableError is an error returned by the server which may be retried at a later time.
+type retriableError struct{ err error }
 
-		age := time.Since(payload.creationDate)
+// Error implements error.
+func (e retriableError) Error() string { return e.err.Error() }
 
-		// Payloads are kept in order so as soon as we find one that isn't, we can break out
-		if age < s.conf.MaxAge {
-			break
-		}
+const (
+	headerAPIKey    = "DD-Api-Key"
+	headerUserAgent = "User-Agent"
+)
 
-		err := fmt.Errorf("payload is older than max age: age=%v, max age=%v", age, s.conf.MaxAge)
-		log.Tracef("Discarding payload: err=%v, payload=%v", err, payload)
-		s.notifyError(payload, err)
-		next = s.removeQueuedPayload(e)
+func (s *sender) do(req *http.Request) error {
+	req.Header.Set(headerAPIKey, s.cfg.apiKey)
+	req.Header.Set(headerUserAgent, userAgent)
+	resp, err := s.cfg.client.Do(req)
+	if err != nil {
+		// request errors include timeouts or name resolution errors and
+		// should thus be retried.
+		return &retriableError{err}
 	}
-}
-
-// Payloads are kept in order so dropping the one at the front guarantees we're dropping the oldest
-func (s *queuableSender) dropOldestPayload(reason string) (*payload, error) {
-	if s.queuedPayloads.Len() == 0 {
-		return nil, fmt.Errorf("no queued payloads")
+	if resp.StatusCode/100 == 5 {
+		// 5xx errors can be retried
+		return &retriableError{
+			fmt.Errorf("server responded with %q", resp.Status),
+		}
+	}
+	if resp.StatusCode/100 != 2 {
+		// status codes that are neither 2xx nor 5xx are considered
+		// non-retriable failures
+		return errors.New(resp.Status)
 	}
+	return nil
+}
 
-	err := fmt.Errorf("payload dropped: %s", reason)
-	droppedPayload := s.queuedPayloads.Front().Value.(*payload)
-	s.removeQueuedPayload(s.queuedPayloads.Front())
-	s.notifyError(droppedPayload, err)
+// payloads specifies a payload to be sent by the sender.
+type payload struct {
+	body    *bytes.Buffer     // request body
+	headers map[string]string // request headers
+}
 
-	return droppedPayload, nil
+// ppool is a pool of payloads.
+var ppool = &sync.Pool{
```

Comment:
Cool, was wondering what this was for. `Pool's purpose is to cache allocated but unused items for later reuse, relieving pressure on the garbage collector.`